File: 2015/01/cs/Program.cs
using System;
using System.Collections.Generic;
using static System.Console;
using System.Linq;
using System.IO;
using System.Diagnostics;

namespace AoC
{
    class Program
    {
        static int Part2(IEnumerable<int> directions)
        {
            var currentFloor = 0;
            foreach (var (direction, index) in directions.Select((direction, index) => (direction, index)))
            {
                currentFloor += direction;
                if (currentFloor == -1)
                    return index + 1;
            }
            throw new Exception("Did not go below 0!");
        }

        static (int, int) Solve(IEnumerable<int> directions)
            => (directions.Sum(), Part2(directions));

        static IEnumerable<int> GetInput(string filePath)
            => !File.Exists(filePath) ? throw new FileNotFoundException(filePath)
            : File.ReadAllText(filePath).Trim().Select(c => c == '(' ? 1 : -1);

        static void Main(string[] args)
        {
            if (args.Length != 1) throw new Exception("Please, add input file path as parameter");

            var watch = Stopwatch.StartNew();
            var (part1Result, part2Result) = Solve(GetInput(args[0]));
            watch.Stop();
            WriteLine($"P1: {part1Result}");
            WriteLine($"P2: {part2Result}");
            WriteLine();
            WriteLine($"Time: {(double)watch.ElapsedTicks / 100 / TimeSpan.TicksPerSecond:f7}");
        }
    }
}

File: 2015/01/scala/Program.scala
import scala.util.control.Breaks._

type Input = List[Int]
type Result = (Int, Int)

def part2(directions: Input): Int =
  var currentFloor = 0
  var result = -1
  breakable { 
    for index <- directions.indices do
      currentFloor += directions(index)
      if currentFloor == -1 then
        result = index + 1
        break
  }
  if result == -1 then
    throw new Exception("Did not go below 0!")
  result

def solve(directions: Input): Result =
  (directions.sum, part2(directions))

def getInput(filePath: String): Input =
  val text = scala.io.Source.fromFile(filePath).mkString
  val directions = for direction <- text yield direction match
    case '(' => 1
    case ')' => -1
    case _ => throw new Exception(s"Uknown direction $direction")
  directions.toList

def main(args: Array[String]) = 
  if args.size != 1 then
    throw new Exception("Please, add input file as parameter")
  val input = getInput(args(0))
  var start = System.nanoTime
  val (part1, part2) = solve(input)
  var end = System.nanoTime
  println(s"Part 1: $part1")
  println(s"Part 2: $part2")
  println(f"\nTime: ${(end - start) * 1e-9}%.7f")
